### PR TITLE
Native Support for Adventure Library

### PR DESCRIPTION
**Pull request by zvyap**
This pull request addresses issue #29, which I opened previously, to add native support for the Adventure API Component.

### Changes in this PR

* Added a new module: `configlib-adventure`, shaded into the Paper and Velocity modules (since both natively support Adventure).
* Implemented serializers for `Component`, `Key`, and `Sound`.
* Added related tests (some were initially AI-generated, but I reviewed, modified, and approved them).
* Added related Javadocs.
* Updated readme

### Notes

* I noticed that running `gradle test` at the root or within any module only executed tests in the `core` module. I am not sure whether this behavior was intentional.
  I modified the behavior so that:

  * Running tests at the root executes all tests across all modules.
  * Running tests inside a specific library module runs only that module’s tests.
  * Running tests inside a plugin module continues to run core tests, as per the previous behavior.

  If the previous test configuration was intentional, please let me know and I will revert the changes accordingly.

* I also fixed two failing tests in the `core` module that were caused by Windows-specific behavior (as shown in the screenshot). All core module tests now pass on Windows.

* If you are not satisfied with the changes mentioned above, I am willing to revert or adjust them as needed.

### CODE REVIEW

**Comment by Emibergo02** [comment]
This is cool!

**Comment by Exlll** [comment]
Hey, thanks a lot for your contribution! I have a lot of stuff going on in my private life lately but I will try to find some time this weekend to have a look at it.

File: configlib-adventure/src/main/java/de/exlll/configlib/ComponentFormat.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    /** Minecraft JSON format for components. */
+    MINECRAFT_JSON(input -> {
+        input = input.trim();
+        return input.startsWith("{") && input.endsWith("}");
+    });
+
+    // Hack to avoid compiler error while singleton pattern initialization
+    private static class Patterns {
+        // Pattern to detect any <tag> in a string
+        static final Pattern MINI_MESSAGE_PATTERN = Pattern.compile("<[^>]+>");
```
Comment (Exlll) [inline configlib-adventure/src/main/java/de/exlll/configlib/ComponentFormat.java]: This pattern is overly broad and matches strings like `< gre eeen >` and `3 < 4 and 4 > 2`.

File: configlib-adventure/src/main/java/de/exlll/configlib/ComponentFormat.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+
+/**
+ * Represents the different text formats supported for Adventure Component serialization.
+ */
+public enum ComponentFormat {
+    /** MiniMessage format with tags like {@code <red>} or {@code <bold>}. */
+    MINI_MESSAGE(input -> Patterns.MINI_MESSAGE_PATTERN.matcher(input).find()),
```
Comment (Exlll) [inline configlib-adventure/src/main/java/de/exlll/configlib/ComponentFormat.java]: You can just use `Pattern.compile("...").asPredicate()`

File: configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        this.deserializeOrder = List.copyOf(deserializeOrder);
+    }
+
+    /**
+     * Creates a new ComponentSerializer using the same format order for both serialization and deserialization.
+     *
+     * @param formats the formats to use, in order of preference
+     */
+    public ComponentSerializer(ComponentFormat... formats) {
+        this(Arrays.asList(formats), Arrays.asList(formats));
```
Comment (Exlll) [inline configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java]: You should make a defensive copy here, too. Calling `asList` is not enough.

File: configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            return null;
+        }
+
+        for (ComponentFormat format : serializeOrder) {
+            try {
+                return serialize(element, format);
+            } catch (Exception ignored) {
+                // This should never happen, but in case of adventure library issue
+                ignored.printStackTrace();
+            }
```
Comment (Exlll) [inline configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java]: Don't do this.

File: configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            if (!format.matches(element)) {
+                continue;
+            }
+
+            try {
+                return deserialize(element, format);
+            } catch (Exception ignored) {
+                // This should never happen, but in case of adventure library issue
+                ignored.printStackTrace();
+            }
```
Comment (Exlll) [inline configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java]: Don't do this.

File: configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    private String serialize(Component component, ComponentFormat format) {
+        return switch (format) {
+            case MINI_MESSAGE -> MiniMessage.miniMessage().serialize(component);
+            case LEGACY_AMPERSAND -> LegacyComponentSerializer.legacyAmpersand().serialize(component);
+            case LEGACY_SECTION -> LegacyComponentSerializer.legacySection().serialize(component);
+            case MINECRAFT_JSON -> GsonComponentSerializer.gson().serialize(component);
+            case TRANSLATION_KEY -> PlainTextComponentSerializer.plainText().serialize(component);
+            default -> throw new UnsupportedOperationException("Unsupported format for serialization: " + format);
```
Comment (Exlll) [inline configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java]: Redundant `default`, please remove.

File: configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    private Component deserialize(String string, ComponentFormat format) {
+        return switch (format) {
+            case MINI_MESSAGE -> MiniMessage.miniMessage().deserialize(string);
+            case LEGACY_AMPERSAND -> LegacyComponentSerializer.legacyAmpersand().deserialize(string);
+            case LEGACY_SECTION -> LegacyComponentSerializer.legacySection().deserialize(string);
+            case MINECRAFT_JSON -> GsonComponentSerializer.gson().deserialize(string);
+            case TRANSLATION_KEY -> Component.translatable(string);
+            default -> throw new UnsupportedOperationException("Unsupported format for deserialization: " + format);
```
Comment (Exlll) [inline configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java]: Redundant `default`, please remove.

File: configlib-adventure/src/test/java/de/exlll/configlib/KeySerializerTest.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        Key key = serializer.deserialize("value");
+        assertEquals(Key.key("default", "value"), key);
+    }
+
+    @Test
+    void testDeserializeWithColonAndDefaultNamespace() {
+        KeySerializer serializer = new KeySerializer("default");
+        // Key.key("default", "namespace:value") will throw InvalidKeyException because
+        // ':' is not allowed in value
+        assertThrows(Exception.class, () -> serializer.deserialize("namespace:value"));
```
Comment (Exlll) [inline configlib-adventure/src/test/java/de/exlll/configlib/KeySerializerTest.java]: Don't expect `Exception` but `InvalidKeyException` in this case

File: configlib-core/src/test/java/de/exlll/configlib/SerializersTest.java [hunk 9]
```diff
@@ -1275,7 +1277,7 @@ void deserializeObjectToStringIfDeserializationTypeAdded() {
 
             assertThat(
                     serializer.deserialize(new File(TMP_CONFIG_PATH)),
-                    is("/tmp/config.yml")
+                    is(TestUtils.isWindows() ? "C:\\tmp\\config.yml" : "/tmp/config.yml")
```
Comment (Exlll) [inline configlib-core/src/test/java/de/exlll/configlib/SerializersTest.java]: Thanks for fixing this. You can use `TestUtils.createPlatformSpecificFilePath` here.

File: configlib-velocity/src/main/java/de/exlll/configlib/ConfigLib.java [hunk 10]
```diff
@@ -3,14 +3,17 @@
...
-        id = "configlib",
-        name = "ConfigLib",
-        version = "4.8.0",
-        url = "https://github.com/Exlll/ConfigLib",
-        description = "A library for working with YAML configurations.",
-        authors = {"Exlll"}
-)
-public final class ConfigLib {}
+@Plugin(id = "configlib", name = "ConfigLib", version = "4.8.0", url = "https://github.com/Exlll/ConfigLib", description = "A library for working with YAML configurations.", authors = {
+        "Exlll" })
```
Comment (Exlll) [inline configlib-velocity/src/main/java/de/exlll/configlib/ConfigLib.java]: Please restore the original format.

File: configlib-adventure/src/main/java/de/exlll/configlib/SoundSerializer.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+     * @param defaultSource the default sound source to use when deserializing
+     */
+    public SoundSerializer(Sound.Source defaultSource) {
+        this.defaultSource = defaultSource;
+    }
+
+    @Override
+    public String serialize(Sound element) {
+        StringBuilder builder = new StringBuilder(element.name().asString());
+        if (element.source() != null) {
```
Comment (Exlll) [inline configlib-adventure/src/main/java/de/exlll/configlib/SoundSerializer.java]: This if statement is always true.

File: configlib-adventure/src/main/java/de/exlll/configlib/SoundSerializer.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+                if (!hasSecondFloat) {
+                    pitch = lastFloat;
+                    endIndex--;
+                }
+            }
+        }
+
+        Key key = buildKey(parts, endIndex);
+        return Sound.sound(key, source, volume, pitch);
+    }
```
Comment (Exlll) [inline configlib-adventure/src/main/java/de/exlll/configlib/SoundSerializer.java]: This parsing logic seems overly complex. Perhaps you can replace this with a `Pattern`? Something like this:

```java
Pattern pattern = Pattern.compile(
        """
        ^(?<key>[a-zA-Z:]+)\
        (?:\\s+(?<volume>\\d+(?:\\.\\d+)?)?)?\
        (?:\\s+(?<pitch>\\d+(?:\\.\\d+)?)?)?\
        (?:\\s+(?<source>MASTER|AMBIENT|MUSIC)?)?\
        \\s*$\
        """);
Matcher matcher = pattern.matcher("minecraft 0.4 MASTER");
System.out.println(matcher.matches());
System.out.println(matcher.group("key"));
System.out.println(matcher.group("volume"));
System.out.println(matcher.group("pitch"));
System.out.println(matcher.group("source"));
```

**Review by Exlll — CHANGES_REQUESTED** [review]
Besides the things that I commented: I generally try to stay below 80 chars per line. 90 chars per line is a hard limit except for some edge cases and in tests. Could you please reformat your code so that the line length is within these limits?
Also please prefix all top-level classes with `Adventure`, i.e. `AdventureComponentSerializer`, `AdventureComponentFormat`, etc.
Overall pretty nice work 👍

File: configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java (hunk 3 above)
Comment (zvyap) [reply configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java]: I already called List.copyOf in another constructor. This constructor just converts the array to a List to pass it to the other constructor. or I miss understanding here?

File: configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java (hunk 4 above)
Comment (zvyap) [reply configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java]: Removed, if runtime exception occurs, it will be handled by serializer caller

File: configlib-adventure/src/main/java/de/exlll/configlib/SoundSerializer.java (hunk 11 above)
Comment (zvyap) [reply configlib-adventure/src/main/java/de/exlll/configlib/SoundSerializer.java]: Fixed by `if (element.source() != defaultSource) {` instead of `null`

File: configlib-adventure/src/main/java/de/exlll/configlib/SoundSerializer.java (hunk 12 above)
Comment (zvyap) [reply configlib-adventure/src/main/java/de/exlll/configlib/SoundSerializer.java]: Fixed by using regex (provide by you) with dynamic generation to respect `DELIMINATOR` field and `Sound.Source` enum

Comment (zvyap) [reply configlib-adventure/src/main/java/de/exlll/configlib/SoundSerializer.java]: Should I fallback to defaultSource if input source is not valid or throw an Exception? The current way is fallback to defaultSource.

File: configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java (hunk 7 above)
Comment (zvyap) [reply configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java]: It is my habit to always have a default for enum switch case to avoid future enum expansion causing issues, but it should be fine in this project since it contains unit test.

Removed.

File: configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java (hunk 6 above)
Comment (zvyap) [reply configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java]: It is my habit to always have a default for enum switch case to avoid future enum expansion causing issues, but it should be fine in this project since it contains unit test.

Removed.

File: configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java (hunk 5 above)
Comment (zvyap) [reply configlib-adventure/src/main/java/de/exlll/configlib/ComponentSerializer.java]: Removed, if runtime exception occurs, it will be handled by serializer caller.

File: configlib-adventure/src/main/java/de/exlll/configlib/ComponentFormat.java (hunk 1 above)
Comment (zvyap) [reply configlib-adventure/src/main/java/de/exlll/configlib/ComponentFormat.java]: Fixed by replace regex to a strict version. `<[a-zA-Z0-9_:-]+(?::[^<>]+)?>`

File: configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentFormat.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+public enum AdventureComponentFormat {
+    /**
+     * MiniMessage format with tags like {@code <red>} or {@code <bold>}.
+     */
+    MINI_MESSAGE(Patterns.MINI_MESSAGE_PATTERN.asPredicate()),
+    /**
+     * Translation key format for translatable components.
+     */
+    TRANSLATION_KEY(input -> true), // translation keys can be any format
+    /**
```
[the remark this answers is not in the document]
Comment (zvyap) [reply configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentFormat.java:21]: On propose, not enable by default. Requires manual setup by advanced users for their own use.

File: configlib-adventure/build.gradle.kts [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+val adventureVersion = "4.26.1"
+
+dependencies {
+    api(project(":configlib-core"))
+    compileOnly("net.kyori:adventure-api:$adventureVersion")
+    compileOnly("net.kyori:adventure-text-minimessage:$adventureVersion")
+    compileOnly("net.kyori:adventure-text-serializer-legacy:$adventureVersion")
+    compileOnly("net.kyori:adventure-text-serializer-gson:$adventureVersion")
+    compileOnly("net.kyori:adventure-text-serializer-plain:${adventureVersion}")
```
[the remark this answers is not in the document]
Comment (zvyap) [reply configlib-adventure/build.gradle.kts:14]: Try to avoid `api / implementation / compileOnlyApi` to avoid double Adventure dependency on user project

File: configlib-adventure/src/test/java/de/exlll/configlib/AdventureKeySerializerTest.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Test
+    void testDeserializeWithColonAndDefaultNamespace() {
+        AdventureKeySerializer serializer = new AdventureKeySerializer("default");
+        // Key.key("default", "namespace:value") will throw InvalidKeyException because
+        // ':' is not allowed in value
+        assertThrows(InvalidKeyException.class, () ->
+                serializer.deserialize("namespace:value"));
+    }
```
[the remark this answers is not in the document]
Comment (zvyap) [reply configlib-adventure/src/test/java/de/exlll/configlib/AdventureKeySerializerTest.java:48]: It does not how it work, `only applies when the input has no namespace` related logic is handle by Adventure library

**Comment by zvyap** [comment]
> Besides the things that I commented: I generally try to stay below 80 chars per line. 90 chars per line is a hard limit except for some edge cases and in tests. Could you please reformat your code so that the line length is within these limits? Also please prefix all top-level classes with `Adventure`, i.e. `AdventureComponentSerializer`, `AdventureComponentFormat`, etc. Overall pretty nice work 👍

- Reformatted code to 80 chars per line, 90 chars hard limit
- Prefixed Adventure related top level classes
- Changes completed except question waiting for reply
- Reviewed and make changes accordingly with suggestion from copilot

@Exlll

File: configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java (hunk 3 above)
Comment (Exlll) [reply configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java]: Yes, you are right, sorry.

File: configlib-adventure/src/main/java/de/exlll/configlib/SoundSerializer.java (hunk 12 above)
Comment (Exlll) [reply configlib-adventure/src/main/java/de/exlll/configlib/SoundSerializer.java]: I think throwing a `ConfigurationException` would be better because it is a configuration error that users should be made aware of. What do you think?

**Comment by Exlll** [comment]
@zvyap Thanks a lot for implementing all the changes! I'll make some room these days to review and merge the PR.

File: configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Override
+    public String serialize(Component element) {
+        if (element == null) {
+            return null;
+        }
+
+        for (AdventureComponentFormat format : serializeOrder) {
+            return serialize(element, format);
+        }
```
Comment (Exlll) [inline configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java]: Hey, so after the change, this for-loop doesn't really make sense, does it? If the `serializeOrder` list is not empty, then its first element is always chosen. Therefore, the first parameter of the constructor of this class could be of `AdventureComponentFormat` type instead of being a list.

File: configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+            return null;
+        }
+
+        for (AdventureComponentFormat format : deserializeOrder) {
+            if (!format.matches(element)) {
+                continue;
+            }
+
+            return deserialize(element, format);
+        }
```
Comment (Exlll) [inline configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java:70]: Does it really make sense to allow different deserialization types? I think trying deserialization with the same format that was used for serialization (and if that fails falling back to MiniMessage) would be more intuitive and also make the API simpler:

```java
public AdventureComponentSerializer(AdventureComponentFormat format) {
    this.format = format;
}
```

I don't yet see why anyone would want to allow their users to mix different formatting styles for messages. Do you personally have a use-case for which you need multiple formats for deserialization? If not, I think we should stick with the simpler implementation until someone requests the more complex one.

Comment (zvyap) [reply configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java:70]: There are **three** reasons why I choose to implement multiple serializer formats:

1. Most people (especially advanced users who have memorized all the color codes) still use `&` or `§` because it is easier, faster, and more efficient to type. However, in some use cases, there are scenarios that are only supported by MiniMessage, such as hover events, click-to-run command messages, or tags like `<rainbow>`.
2. Some plugins may be upgraded from older versions (without Adventure) to newer versions (with Adventure). In such cases, the existing configuration files need to remain compatible with the old format. If the serializer can only resolve one format, it would limit plugin authors to either stay with the old format (`&` or `§`) or completely migrate and break backward compatibility. By supporting multiple formats, plugin authors can gradually introduce MiniMessage (and potentially deprecate the old format) while still maintaining compatibility with existing configurations.
3. Supporting multiple serialization formats does not negatively affect users who prefer to use only a single format. Users who require multiple formats, as well as those who only need one format, can use the library API without significant differences in usage.

Reason 1 is also the main reason why I choose to use library to deserialize Component object from my config instead of manually did it.

File: configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java (hunk 16 above)
Comment (zvyap) [reply configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java]: Yes, sorry. This was my logical mistake during the implementation. I think I will just add a constructor parameter to define the deserialize format instead of reusing the serializeOrder list.

I will implement the change when I free.

File: configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java (hunk 17 above)
Comment (Exlll) [reply configlib-adventure/src/main/java/de/exlll/configlib/AdventureComponentSerializer.java:70]: Gotcha, thanks for the explanation. Then we can leave this paramater as a list.

File: configlib-adventure/src/main/java/de/exlll/configlib/SoundSerializer.java (hunk 12 above)
Comment (zvyap) [reply configlib-adventure/src/main/java/de/exlll/configlib/SoundSerializer.java]: Your point make sense, implemented

**Review by Exlll — APPROVED** [review]
